File: backend/matching-engine/src/main/java/com/commonsengine/matching/infrastructure/persistence/WorkerLocationRepository.java
package com.commonsengine.matching.infrastructure.persistence;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

/**
 * 劳动者位置 JPA Repository——含边界框空间查询。
 */
public interface WorkerLocationRepository extends JpaRepository<WorkerLocationEntity, String> {

    Optional<WorkerLocationEntity> findByWorkerId(String workerId);

    /**
     * 边界框查询——在 (minLat, minLng) ~ (maxLat, maxLng) 矩形内的劳动者。
     *
     * 利用 lat/lng 列上的 B-tree 索引快速过滤，再由应用层用 Haversine 精筛。
     * 比原生 PostGIS ST_Within 轻量，适合 MVP 阶段。
     */
    @Query("SELECT w FROM WorkerLocationEntity w " +
            "WHERE w.lat BETWEEN :minLat AND :maxLat " +
            "AND w.lng BETWEEN :minLng AND :maxLng")
    List<WorkerLocationEntity> findInBoundingBox(
            @Param("minLat") double minLat,
            @Param("maxLat") double maxLat,
            @Param("minLng") double minLng,
            @Param("maxLng") double maxLng
    );
}

File: backend/matching-engine/src/main/kotlin/com/commonsengine/matching/infrastructure/persistence/WorkerLocationRepository.kt
package com.commonsengine.matching.infrastructure.persistence

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

/**
 * 劳动者位置 Repository
 *
 * 空间检索使用 bounding box 预筛选（DB 无关，H2 兼容）。
 * 生产环境 PostgreSQL+PostGIS 可替换为 ST_DWithin 原生查询获得更高性能。
 */
@Repository
interface WorkerLocationRepository : JpaRepository<WorkerLocationEntity, String> {

    /**
     * Bounding box 查询：返回矩形区域内的活跃劳动者。
     *
     * 矩形由 (minLat, maxLat, minLng, maxLng) 定义，
     * 是圆形搜索半径的外接矩形——快速排除绝大部分不相关记录。
     *
     * 排除满单劳动者（active_order_count < maxActiveOrders）。
     */
    @Query("""
        SELECT w FROM WorkerLocationEntity w
        WHERE w.lat BETWEEN :minLat AND :maxLat
          AND w.lng BETWEEN :minLng AND :maxLng
          AND w.activeOrderCount < :maxActiveOrders
        ORDER BY w.rating DESC
    """)
    fun findInBoundingBox(
        @Param("minLat") minLat: Double,
        @Param("maxLat") maxLat: Double,
        @Param("minLng") minLng: Double,
        @Param("maxLng") maxLng: Double,
        @Param("maxActiveOrders") maxActiveOrders: Int,
    ): List<WorkerLocationEntity>

    /**
     * 更新劳动者位置（心跳上报）
     */
    @Modifying
    @Query("""
        UPDATE WorkerLocationEntity w
        SET w.lat = :lat, w.lng = :lng, w.lastSeenAt = :timestamp
        WHERE w.workerId = :workerId
    """)
    fun updateLocation(
        @Param("workerId") workerId: String,
        @Param("lat") lat: Double,
        @Param("lng") lng: Double,
        @Param("timestamp") timestamp: java.time.Instant,
    ): Int
}
